walkeeper/src/safekeeper.rs

```diff
@@ -0,0 +1,444 @@
+    pub wal_seg_size: u32,
+}
+
+/// Acceptor -> Proposer initial response: the highest term known to me
+/// (acceptor voted for).
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AcceptorGreeting {
+    term: u64,
+}
+
+/// Vote request sent from proposer to safekeepers
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VoteRequest {
+    term: Term,
+}
+
+/// Vote itself, sent from safekeeper to proposer
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VoteResponse {
+    term: Term,      // not really needed, just a sanity check
+    vote_given: u64, // fixme u64 due to padding
+    /// Safekeeper's log position, to let proposer choose the most advanced one
+    epoch: Term,
+    flush_lsn: Lsn,
+    restart_lsn: Lsn,
+}
+
+/// Request with WAL message sent from proposer to safekeeper. Along the way it
+/// announces 1) successfull election (with VCL); 2) commit_lsn.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AppendRequest {
+    h: AppendRequestHeader,
+    wal_data: Bytes,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AppendRequestHeader {
+    term: Term,
+    /// volume commit LSN
+    vcl: Lsn,
+    /// start position of message in WAL
+    begin_lsn: Lsn,
+    /// end position of message in WAL
+    end_lsn: Lsn,
+    /// LSN committed by quorum of safekeepers
+    commit_lsn: Lsn,
+    /// restart LSN position  (minimal LSN which may be needed by proposer to perform recovery)
+    restart_lsn: Lsn,
+    // only for logging/debugging
+    proposer_uuid: PgUuid,
+}
+
+/// Report safekeeper state to proposer
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct AppendResponse {
+    // Current term of the safekeeper; if it is higher than proposer's, the
+    // compute is out of date.
+    pub term: Term,
+    pub epoch: Term,
+    // NOTE: this is physical end of wal on safekeeper; currently it doesn't
+    // make much sense without taking epoch into account, as history can be
+    // diverged.
+    pub flush_lsn: Lsn,
+    pub hs_feedback: HotStandbyFeedback,
+}
+
+/// Proposer -> Acceptor messages
+#[derive(Debug)]
+pub enum ProposerAcceptorMessage {
+    Greeting(ProposerGreeting),
+    VoteRequest(VoteRequest),
+    AppendRequest(AppendRequest),
+}
+
+fn des_stream_le<T: LeSer, S: io::Read>(stream: &mut S) -> Result<T> {
+    // As the trait bound implies, this always encodes little-endian.
+    Ok(T::des_from(stream)?)
+}
+
+impl ProposerAcceptorMessage {
+    /// Parse proposer message.
+    pub fn parse(msg: Bytes) -> Result<ProposerAcceptorMessage> {
+        // xxx this is not efficient but easy to work with bin_ser
+        let mut stream = msg.reader();
+        // u64 is here to avoid padding; it will be removed once we stop packing C structs into the wire as is
+        let tag = stream.read_u64::<LittleEndian>()? as u8 as char;
+        match tag {
+            'g' => {
+                let msg = des_stream_le::<ProposerGreeting, Reader<Bytes>>(&mut stream)?;
```

Comment:
This could be simply:
```
let msg = ProposerGreeting::des_from(&mut stream)?;
```
The `des_stream_le` function seems to just make these calls more complicated.

Reply:
Hm, indeed.